pd.read_csv rejects int and float dict values

## Describe the bug

`pandas-stubs` rejects `int` and `float` as values in the `dtype` mapping passed to `pd.read_csv`, even though both are accepted by pandas at runtime and documented as valid.

## To Reproduce

```python
from io import StringIO

import pandas as pd

csv = StringIO("i,f\n1,2.5\n")

my_col_types = {
    "i": int,
    "f": float,
}

pd.read_csv(csv, dtype=my_col_types)
```
running
```
mypy.exe --check-untyped-defs mypy-read_csv.py 
```
produces
> mypy-read_csv.py:12: error: Argument "dtype" to "read_csv" has incompatible type "dict[str, type]"; expected "ExtensionDtype | str | dtype[generic[Any]] | type[complex] | type[bool] | type[object] | type[str] | Mapping[Hashable, ExtensionDtype | str | dtype[generic[Any]] | type[complex] | type[bool] | type[object] | type[str]] | Mapping[Hashable, ExtensionDtype | str | dtype[generic[Any]] | type[complex] | type[bool] | type[object] | type[str] | Mapping[Hashable, ExtensionDtype | str | dtype[generic[Any]] | type[complex] | type[bool] | type[object] | type[str]]] | None"  [arg-type]

### Expected behavior

The code should type-check successfully.

Both `int` and `float` are accepted by `pd.read_csv` at runtime:

```python
>>> df = pd.read_csv(csv, dtype={"i": int, "f": float})
>>> df.dtypes
i      int64
f    float64
dtype: object
```

### Observed stub inconsistency

The current stub allows:

- `type[str]`
- `type[bool]`
- `type[object]`
- `type[complex]`

but not:

- `type[int]`
- `type[float]`

This appears inconsistent with the runtime API.

### Curiously...

When I _inline_ `my_col_types` (i.e., use the literal `{...}` in the `pd.read_csv` call), the error disappears.

## System specifications

- Windows 11 10.0.26100
- python version: 3.14.6
- mypy 2.3.0 (compiled: yes)
- version of installed `pandas-stubs`: 3.0.3.260530

--- a/pandas-stubs/io/parsers/readers.pyi
+++ b/pandas-stubs/io/parsers/readers.pyi
@@ -25,6 +25,7 @@ from pandas._typing import (
     DtypeBackend,
     FilePath,
     HashableT,
+    HashableT1,
     ListLikeHashable,
     ReadCsvBuffer,
     StorageOptions,
@@ -41,7 +42,7 @@ def read_csv(
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = None,
     usecols: UsecolsArgType[HashableT] = None,
-    dtype: DtypeArg | Mapping[Hashable, DtypeArg] | None = None,
+    dtype: DtypeArg | Mapping[HashableT1, DtypeArg] | None = None,
     engine: CSVEngine | None = None,
     converters: (
         Mapping[int | str, Callable[[str], Any]]
@@ -96,7 +97,7 @@ def read_csv(
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = None,
     usecols: UsecolsArgType[HashableT] = None,
-    dtype: DtypeArg | Mapping[Hashable, DtypeArg] | None = None,
+    dtype: DtypeArg | Mapping[HashableT1, DtypeArg] | None = None,
     engine: CSVEngine | None = None,
     converters: (
         Mapping[int | str, Callable[[str], Any]]
@@ -151,7 +152,7 @@ def read_csv(
     names: ListLikeHashable | None = ...,
     index_col: int | str | Sequence[str | int] | Literal[False] | None = ...,
     usecols: UsecolsArgType[HashableT] = ...,
-    dtype: DtypeArg | Mapping[Hashable, DtypeArg] | None = ...,
+    dtype: DtypeArg | Mapping[HashableT1, DtypeArg] | None = ...,
     engine: CSVEngine | None = ...,
     converters: (
         Mapping[int | str, Callable[[str], Any]]

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -886,6 +886,18 @@ def test_types_read_csv_date(tmp_path: Path) -> None:
     )
 
 
+def test_read_csv_dtype_dict() -> None:
+    # GH 1842
+    csv = io.StringIO("i,f\n1,2.5\n")
+
+    my_col_types = {
+        "i": int,
+        "f": float,
+    }
+
+    check(assert_type(pd.read_csv(csv, dtype=my_col_types), pd.DataFrame), pd.DataFrame)
+
+
 def test_read_table(tmp_path: Path) -> None:
     path_str = str(tmp_path / str(uuid.uuid4()))
     check(assert_type(DF.to_csv(path_str, sep="\t"), None), type(None))

